### ISSUE TITLE
Listing devices cannot connect to server

### ISSUE THREAD

**Issue by Ricky12Awesome** [tomaka/cpal#605]
**CPAL Version**: `0.13.4`
**Distro**: `Garuda Dragonized Linux`

I don't really know what's causing this issue, but here's the code I've tried and the output of it

```rust
use cpal::default_host;
use cpal::traits::{DeviceTrait, HostTrait};

fn main() {
  let host = default_host();

  println!("{:?}", cpal::available_hosts());

  let devices = host
    .devices()
    .map(|it| it.map(|it| it.name()))
    .unwrap()
    .collect::<Result<Vec<_>, _>>()
    .unwrap();

  println!("{:?}", devices);
}
```

```
[Alsa]
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_dmix.c:1035:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:503:(_snd_pcm_usb_stream_open) Unknown field hint
["pipewire", "pulse", "default", "hdmi:CARD=NVidia,DEV=0", "hdmi:CARD=NVidia,DEV=1", "hdmi:CARD=NVidia,DEV=2", "hdmi:CARD=NVidia,DEV=3", "hdmi:CARD=NVidia,DEV=4", "hdmi:CARD=NVidia,DEV=5", "hdmi:CARD=NVidia,DEV=6", "sysdefault:CARD=Generic", "front:CARD=Generic,DEV=0", "surround21:CARD=Generic,DEV=0", "surround40:CARD=Generic,DEV=0", "surround41:CARD=Generic,DEV=0", "surround50:CARD=Generic,DEV=0", "surround51:CARD=Generic,DEV=0", "surround71:CARD=Generic,DEV=0", "iec958:CARD=Generic,DEV=0", "sysdefault:CARD=Microphone", "front:CARD=Microphone,DEV=0", "surround21:CARD=Microphone,DEV=0", "surround40:CARD=Microphone,DEV=0", "surround41:CARD=Microphone,DEV=0", "surround50:CARD=Microphone,DEV=0", "surround51:CARD=Microphone,DEV=0", "surround71:CARD=Microphone,DEV=0", "iec958:CARD=Microphone,DEV=0", "sysdefault:CARD=C920", "front:CARD=C920,DEV=0"]

Process finished with exit code 0
```

**Comment by LeCodingWolfie** [comment]
<sup>~Yay! Too much text nobody is gonna read!~</sup>

**TL:DR (Too Long, Didn't Read):** cpal is not able to handle/catch these errors, as neither[`alsa-rs`][alsa-rs] does, becuase they are written to standard error by ALSA, not Rust; which points to the solution, i.e. to catch stderr. 
Therefore, I will create an issue/pull request to alsa-rs :D

<sup>*P.D. If any other sound library, like JACK, has a similar issue, a similar fix can be written.*</sup>

___

On the other hand, it's worth noting that cpal *ignores* whatever of these could be catched by Rust only, in order to *only* show the available devices, not those who do not work, or error, and yet are known to ALSA (such as JACK audio I/O device, in this case), which is also helpful to know whether an audio device is for input or output, even though it's not an issue.

## Walkthrough

<sup>I couldn't not write this little story, apologies :sweat_smile:</sup>

> Unnecessarily, and by curiosity, I tried to list all devices to select instead of the defaults, and was surprised to see many errors, none which lead to panic or an error in Rust (with Result) that could be fixed.

> That's how, all day, I tried to catch an error within Rust, in my code, cpal and its dependency alsa-rs, until I realized these were thrown by ALSA, nowhere in Rust where they could be catched or suppressed from standard error.

*Without further ado, here's what happens.*

### Explaination

In Linux, unless specified, cpal defaults to ALSA <sup>[1][1], [2][2]</sup>, contrary to what the errors may say [^1]. [`devices`][devices()] has to be called so we can fetch all devices, as it is not done upon creation; however, the method only creates a new [`Devices`][Devices] <sup>[3][3]</sup> instance, which we have to iterate over to get each device ([`next`][next]) <sup>[4][4]</sup>, which is why the error is thrown even when using [`count`][count].

`next` calls `DeviceHandles::open` <sup>[5][5]</sup>, which tries to open a handle for both playback and capture, and returns if at least one was opened successfully <sup>[6][6]</sup>, which is useful to know if a device is either for I/O, even though it's determined differently <sup>[7][7]</sup>, and is better than the current approach.

Anyways, on both directions `try_open` is called with `PCM::new` from `alsa-rs` <sup>[8][8]</sup>, which returns a `Result` from `open` <sup>[9][9]</sup> catched through the macro `acheck!` <sup>[10][10]</sup>, which calls the Rust bindings generated in [`alsa-sys`][alsa-sys] from ALSA.

At the end, the error roots from `acheck!`, which catches a error different to the ones printed, and nowhere suppresses the errors that are printed by ALSA to standard error, which means this is not a cpal issue, and must be fixed in alsa-rs.

### Solution

Before anything, to temporarily edit alsa-rs code unto a project, [`rhack`][rhack] <sup>[11][11]</sup> can be used to readily download the crate code and set it up on the current project, or otherwise manually with Cargo.toml's  `[patch]` <sup>[12][12]</sup> (can set a GitHub repository).

We just have to catch the errors from standard error, which we can either print or not. In fact, a similar issue, which is "how to catch a C library's stderr in Rust" <sup>[13][13]</sup>, has already been solved, and is used as reference in the following code.

<details>
  <summary>Code</summary>

`src/main.rs`:
```rust
let host = cpal::default_host();

println!("{:?}", cpal::available_hosts());

let devices = host.devices()
    .map(|devices| devices.map(|d| d.name())).unwrap()
    .collect::<Result<Vec<String>, cpal::DeviceNameError>>().unwrap();

println!("{:?}", devices);
```

[`alsa-rs/src/pcm.rs`:][9]
```rust
fn catch<F>(mut f: F) -> Result<PCM>
    where F: FnMut() -> Result<PCM> {
    use std::os::unix::io::FromRawFd;
    use std::io::prelude::BufRead;
    use std::io::BufReader;
    use std::fs::File;

    extern "C" {
        fn pipe(fd: *mut i32) -> i32;
        fn close(fd: i32) -> i32;
        fn dup2(old_fd: i32, new_fd: i32) -> i32;
    }

    const PIPE_READ: usize = 0;
    const PIPE_WRITE: usize = 1;
    const STDERR_FILENO: i32 = 2;

    // Creates a pipe of standard error (2) which writes on a file descriptor that can be read on; 
    // returns if unsucessful.
    let mut pipefd = [-1; 2];
    if unsafe { pipe(&mut pipefd[0]) } == -1 {
        eprintln!("Unable to create pipe for stderr"); panic!();
    }

    // Creates a file that reads from the pipe
    let file = unsafe {
        File::from_raw_fd(pipefd[PIPE_READ])
    };

    // Redirects the pipe to the file; returns if unsucessful
    if unsafe { dup2(pipefd[PIPE_WRITE], STDERR_FILENO) } == -1 {
        eprintln!("Unable to redirect pipe to file"); panic!();
    }

    // Runs the code that writes to stderr
    let result = f();

    // Prints the first line of the stderr
    let reader = BufReader::new(file);
    let line = reader.lines().next().unwrap().unwrap();
    if line != "" {
        println!("stderr: {:?}", line);
    }

    // Closes stderr
    unsafe {
        close(pipefd[PIPE_WRITE]);
    };

    result
}

// ...

pub fn open(name: &CStr, dir: Direction, nonblock: bool) -> Result<PCM> {

    // ...
    
    catch(|| {
        let result =
            acheck!(snd_pcm_open(&mut r, name.as_ptr(), stream, flags)).map(|_| PCM(r, cell::Cell::new(false))); eprintln!(""); result
    })
}
```

  <sup>This code is only able to catch the first line of the error (which is cleaner), even though I've tried to catch all.</sup>
  <sup>This can only be executed within Linux, due to the use of glibc/Linux libraries.</sup>
  <sup>The code that creates and prints the file from the pipe can be deleted to not print any error.</sup>
</details>

### *Err?*

If you're curious of how to log all the errors from the code flow, here's the code:

<details>
  <summary>Mo' code</summary>

[`cpal/src/host/alsa/mod.rs`:][6]
```rust
fn open(name: &str) -> Result<Self, alsa::Error> {
    if capture_err.is_some() | playback_err.is_some() {
        println!(">> ALSA playback/capture errors: {:?} {:?} in {name}", playback_err, capture_err);
    }
    if let Some(err) = capture_err.and(playback_err) {
        println!("> An error has occurred on both playback/capture: {} in {name}", err); Err(err)
    } else {
        Ok(handles)
    }
}
```

**Optional**

[`alsa-rs/src/pcm.rs`:][14]
```rust
pub fn new(name: &str, dir: Direction, nonblock: bool) -> Result<PCM> {
    let pcm = Self::open(&CString::new(name).unwrap(), dir, nonblock);
    let error = pcm.as_ref().err();
    if error.is_some() {
	    println!(">>> ALSA PCM error: {:?} in {name}", error);
    }; pcm
}
```

[`cpal/src/host/alsa/enumerate.rs`:][5]
```rust
match DeviceHandles::open(&name) {
    Ok(handles) => return Some(Device {
        name,
        handles: Mutex::new(handles),
    }),
    Err(e) => println!("< ALSA error: {}", e)
}
```
</details>

and here's how it looks (without the optional logs):

<details>
  <summary>Output</summary>

```
[Alsa]
stderr: "Cannot connect to server socket err = No such file or directory"
stderr: "Cannot connect to server socket err = No such file or directory"
>> ALSA playback/capture errors: Some(Error("snd_pcm_open", ENOENT)) Some(Error("snd_pcm_open", ENOENT)) in jack
> An error has occurred on both playback/capture: ALSA function 'snd_pcm_open' failed with error 'ENOENT: No such file or directory' in jack
stderr: "ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp"
stderr: "ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp"
>> ALSA playback/capture errors: Some(Error("snd_pcm_open", ENOENT)) Some(Error("snd_pcm_open", ENOENT)) in oss
> An error has occurred on both playback/capture: ALSA function 'snd_pcm_open' failed with error 'ENOENT: No such file or directory' in oss
stderr: "ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map"
stderr: "ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map"
>> ALSA playback/capture errors: Some(Error("snd_pcm_open", EINVAL)) Some(Error("snd_pcm_open", EINVAL)) in surround21:CARD=PCH,DEV=0
> An error has occurred on both playback/capture: ALSA function 'snd_pcm_open' failed with error 'EINVAL: Invalid argument' in surround21:CARD=PCH,DEV=0
stderr: "ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map"
stderr: "ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map"
>> ALSA playback/capture errors: Some(Error("snd_pcm_open", EINVAL)) Some(Error("snd_pcm_open", EINVAL)) in surround41:CARD=PCH,DEV=0
> An error has occurred on both playback/capture: ALSA function 'snd_pcm_open' failed with error 'EINVAL: Invalid argument' in surround41:CARD=PCH,DEV=0
stderr: "ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map"
stderr: "ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map"
>> ALSA playback/capture errors: Some(Error("snd_pcm_open", EINVAL)) Some(Error("snd_pcm_open", EINVAL)) in surround50:CARD=PCH,DEV=0
> An error has occurred on both playback/capture: ALSA function 'snd_pcm_open' failed with error 'EINVAL: Invalid argument' in surround50:CARD=PCH,DEV=0
>> ALSA playback/capture errors: Some(Error("snd_pcm_open", EACCES)) Some(Error("snd_pcm_open", EACCES)) in usbstream:CARD=PCH
> An error has occurred on both playback/capture: ALSA function 'snd_pcm_open' failed with error 'EACCES: Permission denied' in usbstream:CARD=PCH
>> ALSA playback/capture errors: None Some(Error("snd_pcm_open", ENOENT)) in hdmi:CARD=NVidia,DEV=0
>> ALSA playback/capture errors: None Some(Error("snd_pcm_open", ENOENT)) in hdmi:CARD=NVidia,DEV=1
>> ALSA playback/capture errors: None Some(Error("snd_pcm_open", ENOENT)) in hdmi:CARD=NVidia,DEV=2
>> ALSA playback/capture errors: None Some(Error("snd_pcm_open", ENOENT)) in hdmi:CARD=NVidia,DEV=3
>> ALSA playback/capture errors: Some(Error("snd_pcm_open", EACCES)) Some(Error("snd_pcm_open", EACCES)) in usbstream:CARD=NVidia
> An error has occurred on both playback/capture: ALSA function 'snd_pcm_open' failed with error 'EACCES: Permission denied' in usbstream:CARD=NVidia
["lavrate", "samplerate", "speexrate", "pipewire", "pulse", "speex", "upmix", "vdownmix", "default", "sysdefault:CARD=PCH", "front:CARD=PCH,DEV=0", "surround40:CARD=PCH,DEV=0", "surro
und51:CARD=PCH,DEV=0", "surround71:CARD=PCH,DEV=0", "hdmi:CARD=NVidia,DEV=0", "hdmi:CARD=NVidia,DEV=1", "hdmi:CARD=NVidia,DEV=2", "hdmi:CARD=NVidia,DEV=3"]
```
</details>

In fact, the errors are thrown when cpal cannot open a device with neither playback or capture.
However, it's also possible to catch the individual errors when either one of those do not work, from what I remember.

___

I've learnt some new things, documented the whole process, and *fixed* an issue <sup>(or two)</sup>, didn't hurt, right?

[^1]: As mentioned, this is becuase JACK seems to create an audio device that is detected by ALSA, as shown by output of the errors in [*Err?*](#err?).

[alsa-rs]: https://github.com/diwic/alsa-rs
[alsa-sys]: https://github.com/diwic/alsa-sys
[rhack]: https://github.com/nakabonne/rhack

[devices()]: https://docs.rs/cpal/0.14.0/cpal/traits/trait.HostTrait.html#tymethod.devices
[Devices]: https://docs.rs/cpal/0.14.0/cpal/traits/trait.HostTrait.html#associatedtype.Devices
[count]: https://doc.rust-lang.org/stable/core/iter/trait.Iterator.html#method.count
[next]: https://doc.rust-lang.org/stable/core/iter/trait.Iterator.html#tymethod.next

[11]: https://www.reddit.com/r/rust/comments/urcv5m/comment/i8wjm17/?context=3
[12]: https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section
[13]: https://stackoverflow.com/questions/63600029/can-stdout-stderr-of-a-c-c-library-be-caught-by-rust

[1]: https://github.com/RustAudio/cpal/blob/bd1ec6ae2e6046646e74629cc522f68c2b0ce4ad/src/platform/mod.rs#L594
[2]: https://github.com/RustAudio/cpal/blob/master/examples/record_wav.rs
[3]: https://github.com/RustAudio/cpal/blob/bd1ec6ae2e6046646e74629cc522f68c2b0ce4ad/src/host/alsa/mod.rs#L49
[4]: https://github.com/RustAudio/cpal/blob/bd1ec6ae2e6046646e74629cc522f68c2b0ce4ad/src/host/alsa/enumerate.rs#L25
[5]: https://github.com/RustAudio/cpal/blob/bd1ec6ae2e6046646e74629cc522f68c2b0ce4ad/src/host/alsa/enumerate.rs#L37
[6]: https://github.com/RustAudio/cpal/blob/bd1ec6ae2e6046646e74629cc522f68c2b0ce4ad/src/host/alsa/mod.rs#L188
[7]: https://github.com/RustAudio/cpal/blob/bd1ec6ae2e6046646e74629cc522f68c2b0ce4ad/src/traits.rs#L64
[8]: https://github.com/RustAudio/cpal/blob/bd1ec6ae2e6046646e74629cc522f68c2b0ce4ad/src/host/alsa/mod.rs#L216
[9]: https://github.com/diwic/alsa-rs/blob/a323c994bb7dba52f22e2247ff088ed5194532e2/src/pcm.rs#L124
[10]: https://github.com/diwic/alsa-rs/blob/a323c994bb7dba52f22e2247ff088ed5194532e2/src/error.rs#L22
[14]: https://github.com/diwic/alsa-rs/blob/a323c994bb7dba52f22e2247ff088ed5194532e2/src/pcm.rs#L119

**Comment by dsgallups** [comment]
Forwarding from the issue:

> Okay, so I found it easier to implement myself than to tell someone else how to do it, so now you can call `Output::local_error_handler` to redirect alsa-lib's errors there.
> 
> If you still get errors from jack, that is outside alsa-lib's control unfortunately.


Taking a look at this addition, it's not possible to catch some of the errors. by a one-time construction of `Output::local_error_handler()`, users will be able to capture *some*, but not all errors. Take this code:

```rust
fn main() {
    let outputs = Output::local_error_handler().unwrap();
    let host = cpal::default_host();
    let input_devices = host.input_devices();
    match input_devices {
        Ok(input_devices) => for (_device) in input_devices {},
        Err(e) => {
            eprintln!("Failed to get input audio devices: {}", e);
        }
    }
    let mut errors = String::new();
    let output = outputs.borrow();
    output.buffer_string(|bytes| {
        errors.push_str(str::from_utf8(bytes).unwrap());
    });
    println!("\n\nCaught errors\n{errors}")
}
```

Some errors are caught in the buffer. However, anything caught from jack server is not. Particularly these:
```console
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
```

<details>
  <summary>full console output can be found here</summary>

```console
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock


Caught errors
_snd_pcm_oss_open: Cannot open device /dev/dsp
_snd_pcm_oss_open: Cannot open device /dev/dsp
_snd_pcm_oss_open: Cannot open device /dev/dsp
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dsnoop_open: The dsnoop plugin supports only capture stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dsnoop_open: The dsnoop plugin supports only capture stream
snd_pcm_dmix_open: unable to open slave
snd_pcm_dsnoop_open: unable to open slave
snd_pcm_dsnoop_open: unable to open slave
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dmix_open: The dmix plugin supports only playback stream
snd_pcm_dsnoop_open: The dsnoop plugin supports only capture stream
snd_pcm_dsnoop_open: The dsnoop plugin supports only capture stream
```
</details>

I'm a bit unsure about how to proceed from here. Maybe `alsa` itself needs an update? Or, it might be worthwhile to pipe stderr using the above comment directly in cpal, if possible.
